netmaster/daemon.go

```diff
@@ -0,0 +1,339 @@
+func (d *daemon) networks(id string) ([]core.State, error) {
+	var (
+		err error
+		nw  *mastercfg.CfgNetworkState
+	)
+
+	nw = &mastercfg.CfgNetworkState{}
+	if nw.StateDriver, err = utils.GetStateDriver(); err != nil {
+		return nil, err
+	}
+
+	if id == "all" {
+		return nw.ReadAll()
+	} else if err := nw.Read(id); err == nil {
+		return []core.State{core.State(nw)}, nil
+	}
+
+	return nil, core.Errorf("Unexpected code path")
+}
+
+// runLeader runs leader loop
+func (d *daemon) runLeader() {
+	router := mux.NewRouter()
+
+	// acquire listener mutex
+	d.listenerMutex.Lock()
+	defer d.listenerMutex.Unlock()
+
+	// Create a new api controller
+	d.apiController = objApi.NewAPIController(router)
+
+	// initialize policy manager
+	mastercfg.InitPolicyMgr(d.stateDriver, d.ofnetMaster)
+
+	// Register netmaster service
+	d.registerService()
+
+	// setup HTTP routes
+	d.registerRoutes(router)
+
+	// Create HTTP server and listener
+	server := &http.Server{Handler: router}
+	listener, err := net.Listen("tcp", d.listenURL)
+	if nil != err {
+		log.Fatalln(err)
+	}
+
+	log.Infof("Netmaster listening on %s", d.listenURL)
+
+	// start server
+	go server.Serve(listener)
+
+	// Wait till we are asked to stop
+	<-d.stopLeaderChan
+
+	// Close the listener and exit
+	listener.Close()
+	log.Infof("Exiting Leader mode")
+}
+
+// runFollower runs the follower FSM loop
+func (d *daemon) runFollower() {
+	router := mux.NewRouter()
+	router.PathPrefix("/").HandlerFunc(slaveProxyHandler)
+
+	// acquire listener mutex
+	d.listenerMutex.Lock()
+	defer d.listenerMutex.Unlock()
+
+	// start server
+	server := &http.Server{Handler: router}
+	listener, err := net.Listen("tcp", d.listenURL)
+	if nil != err {
+		log.Fatalln(err)
+	}
+
+	// start server
+	go server.Serve(listener)
+
+	// Register netmaster service
+	d.registerService()
+
+	// just wait on stop channel
+	log.Infof("Listening in follower mode")
+	<-d.stopFollowerChan
+
+	// Close the listener and exit
+	listener.Close()
+	log.Info("Exiting follower mode")
+}
+
+// becomeLeader changes daemon FSM state to master
+func (d *daemon) becomeLeader() {
+	// ask listener to stop
+	d.stopFollowerChan <- true
+
+	// set current state
+	d.currState = "leader"
+
+	// Run the HTTP listener
+	go d.runLeader()
+}
+
+// becomeFollower changes FSM state to follower
+func (d *daemon) becomeFollower() {
+	// ask listener to stop
+	d.stopLeaderChan <- true
+
+	// set current state
+	d.currState = "follower"
+
+	// run follower loop
+	go d.runFollower()
+}
+
+// runMasterFsm runs netmaster FSM
+func (d *daemon) runMasterFsm() {
+	// create new ofnet master
+	d.ofnetMaster = ofnet.NewOfnetMaster(ofnet.OFNET_MASTER_PORT)
+	if d.ofnetMaster == nil {
+		log.Fatalf("Error creating ofnet master")
+	}
+
+	// Create an objdb client
+	d.objdbClient = objdb.NewClient("")
+
+	// Get the address to be used for local communication
+	localIP, err := d.objdbClient.GetLocalAddr()
```

Comment:
This means that the netmasters can run only on the nodes where etcd run in non-proxy mode. If we want to avoid this restriction, we should take a control-ip command line arg like netplugin.